ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java

```diff
@@ -2839,6 +2848,87 @@ public static void setNonTransactional(Map<String, String> tblProps) {
     tblProps.remove(hive_metastoreConstants.TABLE_TRANSACTIONAL_PROPERTIES);
   }
 
+  /**
+   * Look for delta directories matching the list of writeIds and deletes them.
+   * @param rootPartition root partition to look for the delta directories
+   * @param conf configuration
+   * @param writeIds list of writeIds to look for in the delta directories
+   * @return list of deleted directories.
+   * @throws IOException
+   */
+  public static List<FileStatus> deleteDeltaDirectories(Path rootPartition, Configuration conf, Set<Long> writeIds)
+      throws IOException {
+    FileSystem fs = rootPartition.getFileSystem(conf);
+
+    PathFilter filter = (p) -> {
+      String name = p.getName();
+      for (Long wId : writeIds) {
+        if (name.startsWith(deltaSubdir(wId, wId)) && !name.contains("=")) {
+          return true;
+        } else if (name.startsWith(baseDir(wId)) && !name.contains("=")) {
+          return true;
+        }
+      }
+      return false;
+    };
+    List<FileStatus> deleted = new ArrayList<>();
+    deleteDeltaDirectoriesAux(rootPartition, fs, filter, deleted);
+    return deleted;
+  }
+
+  private static void deleteDeltaDirectoriesAux(Path root, FileSystem fs, PathFilter filter, List<FileStatus> deleted)
+      throws IOException {
+    RemoteIterator<FileStatus> it = listIterator(fs, root, null);
+
+    while (it.hasNext()) {
+      FileStatus fStatus = it.next();
+      if (fStatus.isDirectory()) {
+        if (filter.accept(fStatus.getPath())) {
+          fs.delete(fStatus.getPath(), true);
+          deleted.add(fStatus);
+        } else {
+          deleteDeltaDirectoriesAux(fStatus.getPath(), fs, filter, deleted);
+          if (isDirectoryEmpty(fs, fStatus.getPath())) {
+            fs.delete(fStatus.getPath(), false);
+            deleted.add(fStatus);
+          }
+        }
+      }
+    }
+  }
+
+  private static boolean isDirectoryEmpty(FileSystem fs, Path path) throws IOException {
+    RemoteIterator<FileStatus> it = listIterator(fs, path, null);
+    return !it.hasNext();
+  }
+
+  private static RemoteIterator<FileStatus> listIterator(FileSystem fs, Path path, PathFilter filter)
+      throws IOException {
+    try {
+      return new ToFileStatusIterator(SHIMS.listLocatedHdfsStatusIterator(fs, path, filter));
+    } catch (Throwable t) {
```

Comment:
This should be similar to tryListLocatedHdfsStatus don't catch all Throwable

Reply:
removed it